### ISSUE TITLE
to_days(1001-00-00) returns 365576 instead of NULL

### ISSUE THREAD

**Issue by mjonss** [pingcap/tidb#30363]
## Bug Report

to_days(<some invalid date>) does return a value instead of NULL, like MySQL.
This is most likely true for any invalid dates.

Found while testing MySQL test file `partition_pruning.test`

### 1. Minimal reproduce step (Required)

```sql
select to_days('1001-00-00');
```

### 2. What did you expect to see? (Required)

```sql
mysql> select to_days('1001-00-00');
+-----------------------+
| to_days('1001-00-00') |
+-----------------------+
|                  NULL |
+-----------------------+
1 row in set, 1 warning (0,00 sec)

mysql> show warnings;
+---------+------+----------------------------------------+
| Level   | Code | Message                                |
+---------+------+----------------------------------------+
| Warning | 1292 | Incorrect datetime value: '1001-00-00' |
+---------+------+----------------------------------------+
1 row in set (0,00 sec)
```

### 3. What did you see instead (Required)

```sql
tidb> select to_days('1001-00-00');
+-----------------------+
| to_days('1001-00-00') |
+-----------------------+
|                365576 |
+-----------------------+
1 row in set (0,00 sec)
```

### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v5.4.0-alpha-305-gcec4acbf5e
Edition: Community
Git Commit Hash: cec4acbf5ea952a8f92c8258fce3142528f9fb1e
Git Branch: master
UTC Build Time: 2021-12-02 19:58:57
GoVersion: go1.17
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

**Comment by mjonss** [comment]
Also `TO_SECONDS()` is affected the same way.

**Comment by XuHuaiyu** [comment]
There are 2 problems of the issue:
1. According to the MySQL manual: 
> TO_DAYS() is not intended for use with values that precede the advent of the Gregorian calendar (1582), because it does not take into account the days that were lost when the calendar was changed. For dates before 1582 (and possibly a later year in other locales), results from this function are not reliable

2. We do not handle the `zero_date` and `zero_in_date` compatibly for `select-stmt`.

**Comment by XuHuaiyu** [comment]
I think we can fix problem 1 temporarily, and keep the in-compatibility of `zero_date` and `zero_in_date`. Because the `NO_ZERO_DATE` and `NO_ZERO_IN_DATE` sql modes are [deprecated](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_no_zero_date) by MySQL.

**Comment by XuHuaiyu** [comment]
To fix problem 1, we can check the `Gregorian calendar` in https://github.com/pingcap/tidb/blob/28446605c2e019e84e934602f4bd136294a61ff3/expression/builtin_time.go#L6845-L6856

Remember to fix the implementation of vecEvalInt implementation.
https://github.com/pingcap/tidb/blob/28446605c2e019e84e934602f4bd136294a61ff3/expression/builtin_time_vec.go#L1825-L1856

**Comment by fzhedu** [comment]
after discussed with @XuHuaiyu , we return null with warings for invalid date values.

**Comment by fzhedu** [comment]
mysql
```
mysql> insert into tds values (1990-00-01,'1990-00-01');
ERROR 1292 (22007): Incorrect date value: '1989' for column 'dt' at row 1
mysql> select @@sql_mode;
+-----------------------------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                                            |
+-----------------------------------------------------------------------------------------------------------------------+
| ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION |
+-----------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select * from tds;
+------------+------------+
| dt         | ds         |
+------------+------------+
| 2000-10-01 | 1001-00-00 |
+------------+------------+
1 row in set (0.00 sec)

mysql> select to_days(ds), to_days(ds) from tds;
+-------------+-------------+
| to_days(ds) | to_days(ds) |
+-------------+-------------+
|        NULL |        NULL |
+-------------+-------------+
1 row in set, 2 warnings (0.00 sec)

mysql> select to_days(dt), to_days(ds) from tds;
+-------------+-------------+
| to_days(dt) | to_days(ds) |
+-------------+-------------+
|      730759 |        NULL |
+-------------+-------------+
1 row in set, 1 warning (0.01 sec)

mysql> set @@sql_mode="";
Query OK, 0 rows affected (0.00 sec)

mysql> insert into tds values (1990-00-01,'1990-00-01');
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------+
| Level   | Code | Message                                 |
+---------+------+-----------------------------------------+
| Warning | 1265 | Data truncated for column 'dt' at row 1 |
+---------+------+-----------------------------------------+
1 row in set (0.00 sec)

mysql> select to_days(dt), to_days(ds) from tds;
+-------------+-------------+
| to_days(dt) | to_days(ds) |
+-------------+-------------+
|      730759 |        NULL |
|        NULL |        NULL |
+-------------+-------------+
2 rows in set, 2 warnings (0.00 sec)
```
 tidb
```
mysql> select @@sql_mode;
+------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                     |
+------------------------------------------------------------------------------------------------+
| ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO |
+------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql>  create table tds (dt date, ds char(20));
Query OK, 0 rows affected (0.00 sec)

mysql> insert into tds values (1001-00-00,'1001-00-00');
Query OK, 1 row affected (0.01 sec)

mysql> insert into tds values (1990-00-01,'1990-00-01');
ERROR 1292 (22007): Incorrect date value: '1989' for column 'dt' at row 1
mysql> select * from tds;
+------------+------------+
| dt         | ds         |
+------------+------------+
| 2000-10-01 | 1001-00-00 |
+------------+------------+
1 row in set (0.00 sec)

mysql> select to_days(ds), to_days(ds) from tds;
+-------------+-------------+
| to_days(ds) | to_days(ds) |
+-------------+-------------+
|      365576 |      365576 |
+-------------+-------------+
1 row in set (0.00 sec)

mysql> select to_days(dt), to_days(ds) from tds;
+-------------+-------------+
| to_days(dt) | to_days(ds) |
+-------------+-------------+
|      730759 |      365576 |
+-------------+-------------+
1 row in set (0.01 sec)

mysql> set @@sql_mode="";
Query OK, 0 rows affected (0.00 sec)

mysql> insert into tds values (1990-00-01,'1990-00-01');
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql>  select to_days(dt), to_days(ds) from tds;
+-------------+-------------+
| to_days(dt) | to_days(ds) |
+-------------+-------------+
|      730759 |      365576 |
|        NULL |      726802 |
+-------------+-------------+
2 rows in set, 1 warning (0.00 sec)
```

test the sql_mode with write, tidb is compatible with mysql.
the problem comes from the read and compute. so we just check the validibility of value when computing the function.

**Comment by fzhedu** [comment]
root cause
```
mysql> select @@sql_mode;
+-------------------------------------------------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                                                                |
+-------------------------------------------------------------------------------------------------------------------------------------------+
| ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION |
+-------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select cast('1001-00-00' as date);
+----------------------------+
| cast('1001-00-00' as date) |
+----------------------------+
| 1001-00-00                 |
+----------------------------+
1 row in set (0.00 sec)

mysql> set @@sql_mode = "";
Query OK, 0 rows affected (0.00 sec)

mysql> select @@sql_mode;
+------------+
| @@sql_mode |
+------------+
|            |
+------------+
1 row in set (0.00 sec)

mysql> select cast('1001-00-00' as date);
+----------------------------+
| cast('1001-00-00' as date) |
+----------------------------+
| 1001-00-00                 |
+----------------------------+
1 row in set (0.00 sec)
```
the cast does not care the sql_mode, while mysql does.

**Comment by fzhedu** [comment]
mysql
```
+-----------------------------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                                            |
+-----------------------------------------------------------------------------------------------------------------------+
| ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION |
+-----------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select to_days('1001-10-01');
+-----------------------+
| to_days('1001-10-01') |
+-----------------------+
|                365881 |
+-----------------------+
1 row in set (0.00 sec)


mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.26    |
+-----------+
1 row in set (0.01 sec)
```

**Comment by fzhedu** [comment]
mysql does not
```
mysql> set @@sql_mode='';
Query OK, 0 rows affected (0.00 sec)

mysql> select @@sql_mode;
+------------+
| @@sql_mode |
+------------+
|            |
+------------+
1 row in set (0.00 sec)

mysql>  select cast('1001-00-00' as date);
+----------------------------+
| cast('1001-00-00' as date) |
+----------------------------+
| 1001-00-00                 |
+----------------------------+
mysql>  select to_days(cast('1001-00-00' as date));
+-------------------------------------+
| to_days(cast('1001-00-00' as date)) |
+-------------------------------------+
|                                NULL |
+-------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select to_days('1001-00-00');
+-----------------------+
| to_days('1001-00-00') |
+-----------------------+
|                  NULL |
+-----------------------+
1 row in set, 1 warning (0.00 sec)

mysql> set @@sql_mode = "ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION";
Query OK, 0 rows affected (0.00 sec)

mysql> select @@sql_mode;
+-----------------------------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                                            |
+-----------------------------------------------------------------------------------------------------------------------+
| ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION |
+-----------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select to_days('1001-00-00');
+-----------------------+
| to_days('1001-00-00') |
+-----------------------+
|                  NULL |
+-----------------------+
1 row in set, 1 warning (0.00 sec)
mysql>  select to_days(cast('1001-00-00' as date));
+-------------------------------------+
| to_days(cast('1001-00-00' as date)) |
+-------------------------------------+
|                                NULL |
+-------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select to_days('1001-00-00');
+-----------------------+
| to_days('1001-00-00') |
+-----------------------+
|                  NULL |
+-----------------------+
1 row in set, 1 warning (0.00 sec)
```

mysql cast string as date cares the sql_mode, while other functions does not care sql_mode

**Comment by fzhedu** [comment]
mysql 8.0
```
mysql> select month('1001-00-00');
+---------------------+
| month('1001-00-00') |
+---------------------+
|                   0 |
+---------------------+
1 row in set (0.00 sec)

mysql> select day('1001-00-00');
+-------------------+
| day('1001-00-00') |
+-------------------+
|                 0 |
+-------------------+
1 row in set (0.00 sec)

mysql>  select cast('1001-00-00' as date);
+----------------------------+
| cast('1001-00-00' as date) |
+----------------------------+
| NULL                       |
+----------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select to_days('1001-00-00');
+-----------------------+
| to_days('1001-00-00') |
+-----------------------+
|                  NULL |
+-----------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select @@sql_mode;
+-----------------------------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                                            |
+-----------------------------------------------------------------------------------------------------------------------+
| ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION |
+-----------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by solotzg** [comment]
tidb release-6.1
```
MySQL [test]> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                       |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.1.0-nightly
Edition: Community
Git Commit Hash: 1a89decdb192cbdce6a7b0020d71128bc964d30f
Git Branch: heads/refs/tags/v6.1.0-nightly
UTC Build Time: 2022-06-05 10:17:31
GoVersion: go1.18.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)

MySQL [test]> select to_days('1001-00-00');
+-----------------------+
| to_days('1001-00-00') |
+-----------------------+
|                365576 |
+-----------------------+
1 row in set (0.00 sec)
```

tidb master: It has been fixed, but warning message is different with mysql.
```
MySQL [test]> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                                   |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.1.0-nightly
Edition: Community
Git Commit Hash: 173dd005cc345a0be0230ccf6f2eb69a63d82e2d
Git Branch: heads/refs/tags/v6.1.0-nightly
UTC Build Time: 2022-06-03 11:09:29
GoVersion: go1.18.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
Store: tikv |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

MySQL [test]> select to_days('1001-00-00');
+-----------------------+
| to_days('1001-00-00') |
+-----------------------+
|                  NULL |
+-----------------------+
1 row in set, 1 warning (0.00 sec)

MySQL [test]> show warnings;
+---------+------+--------------------------------------------------------+
| Level   | Code | Message                                                |
+---------+------+--------------------------------------------------------+
| Warning | 1292 | Incorrect datetime value: '1001-00-00 00:00:00.000000' |
+---------+------+--------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by zanmato1984** [comment]
Fixed by #34873. Closing.
